src/gluonts/mx/rnn_extend/rnn_cell.py

```diff
@@ -0,0 +1,338 @@
+
+        prev_output = self._prev_output
+        if prev_output is None:
+            prev_output = F.zeros_like(next_output)
+
+        output_mask = mask(p_outputs, next_output)
+        output = (F.where(output_mask, next_output, prev_output)
+                  if p_outputs != 0. else next_output)
+        
+        # only for RNN, the first element of states is output
+        # use the same mask as output, instead of simply copy output to the first element
+        # in case that the base cell is ResidualCell
+        new_states = [F.where(output_mask, next_states[0], states[0])
+                          if p_outputs != 0. else next_states[0]]
+        if self.preserve_raw_output:
+            # the output is stored in the last element of states, thus skip it
+            new_states.extend([F.where(mask(p_states, new_s), new_s, old_s) for new_s, old_s in
+                               zip(next_states[1:], states[1:-1])] if p_states != 0. else next_states[1:])
+            # store raw output
+            new_states.append(next_states[0])
+        else:
+            new_states.extend([F.where(mask(p_states, new_s), new_s, old_s) for new_s, old_s in
+                               zip(next_states[1:], states[1:])] if p_states != 0. else next_states[1:])
+
+        self._prev_output = output
+
+        return output, new_states
+
+
+class AccumulateStatesCell(ModifierCell):
+    """
+    Accumulate part of the states in base cell.
+    To make it work, AccumulateStatesCell must be the last ModifierCell when stacking
+    """
+    def __init__(self, base_cell, index_list):
+        super(AccumulateStatesCell, self).__init__(base_cell)
+        self.index_list = index_list
+
+    def state_info(self, batch_size=0):
+        cell_state_info = self.base_cell.state_info(batch_size)
+        for index in self.index_list:
+            # placeholder for the accumulated states
+            cell_state_info.append({'shape': (-1), '__layout__': 'T', 'state_index': index})
+        return cell_state_info
+
+    def begin_state(self, func=symbol.zeros, **kwargs):
+        assert not self._modified, \
+            "After applying modifier cells the base " \
+            "cell cannot be called directly. Call the modifier cell instead."
+        self.base_cell._modified = False
+        begin = self.base_cell.begin_state(func=func, **kwargs)
+        for index in self.index_list:
+            # placeholder for the accumulated states
+            begin.append([])
+        self.base_cell._modified = True
+        return begin
+
+    def hybrid_forward(self, F, inputs, states):
+        # identity mapping
+        next_output, next_states = self.base_cell(inputs, states)
+        # append accumulated states
+        accumulated_states = []
+        for i, index in enumerate(self.index_list):
+            accumulated_states.append(states[len(states)-len(self.index_list)+i] + [next_states[index]])
```

Comment:
it seems that the first part can be reduced to states[i - len(self.index_list)].